### PR TITLE
Add versioning policy.

### PR DESCRIPTION
**Pull request by znewman01**
Addresses (does not fix) #2518.

This follows [Proposal: Cosign Versioning][versioning-proposal], which in turn follows https://github.com/sigstore/cosign/discussions/2365.

After getting approval on this PR, I will update #2518 to include a checklist containing the following (possibly linking to separate bugs):

- [ ] (docs) Communicate the new version policy to Cosign users.
- [ ] (process/CI) Separate CLI and API releases (with different tagging schemes).
- [ ] (process/CI) Use gorelease or similar to catch breaking API changes.
- [ ] (code) Add library support for deprecations (see also https://github.com/sigstore/cosign/issues/2510)
- [ ] (testing) E2E testing for old Cosign versions (Also client libraries, once they're stable).

[versioning-proposal]: https://docs.google.com/document/d/1urWUPhtzXKWqL9CoaEw4Z35v5IDl9yrTRQ40XlYekOo/edit#

Signed-off-by: Zachary Newman <zjn@chainguard.dev>

#### Release Note

Added a new versioning policy for both the CLI and API.

#### Documentation

Will be addressed in #2518.

### CODE REVIEW

**Comment by znewman01** [comment]
Added all of @sigstore/cosign-codeowners as reviewers. (Maybe we need to do a sweep, I know a few folks aren't actively maintaining anymore.)

**Comment by dlorenc** [comment]
Sweep sounds good separately but we can just hold this for a week or so for any large objections given it's been in review for so long in doc format

**Review by dlorenc — DISMISSED** [review]

File: VERSIONING.md
Comment (hectorj2f) [inline VERSIONING.md:170]: Question: @znewman01 Shouldn't `cmd` be focused on providing the CLI commands without exposing much functionality ? I was expecting to find here `internal/` and `pkg/`.

Comment (znewman01) [reply VERSIONING.md]: That was  awkward wording. From the perspective of Go API consumers, the "public interface" is everything *except* `internal/` and `cmd/` (mostly `pkg/`). My point is that Go API users should not rely on `cmd/`.

I've reworded to:

>  Functionality should be private (either language-private or in a "private"
>  package like `internal/` or `cmd/`) by default, and exposing it publicly
>  will be a deliberate choice.

Hopefully that's clearer.

**Review by priyawadhwa — DISMISSED** [review]
thanks @znewman01 !

File: VERSIONING.md
Comment (loosebazooka) [inline VERSIONING.md:11]: things?

File: VERSIONING.md
Comment (loosebazooka) [inline VERSIONING.md:23]: what `[semantic versioning]` supposed to be a link to something?

File: VERSIONING.md
Comment (loosebazooka) [inline VERSIONING.md:26]: why is there a colon after large changes? `large changes:`

File: VERSIONING.md
Comment (loosebazooka) [inline VERSIONING.md:32]: you can link to "Deprecation" here with `#deprecation` I think?

File: VERSIONING.md
Comment (loosebazooka) [inline VERSIONING.md:38]: did you want this to be a link above? Why not just use inline links? (same question all over this doc)

File: VERSIONING.md
Comment (loosebazooka) [inline VERSIONING.md:46]: was -> is? I dunno, tense feels off

File: VERSIONING.md
Comment (loosebazooka) [inline VERSIONING.md:80]: I dunno if this last really matches this "header" of Unstable Commands?

File: VERSIONING.md
Comment (loosebazooka) [inline VERSIONING.md:109]: Feel like this should be a code block, not a quote?

File: VERSIONING.md
Comment (loosebazooka) [inline VERSIONING.md:121]: lulz "search the codebase"?

File: VERSIONING.md
Comment (znewman01) [reply VERSIONING.md]: lol, "tools/libraries"

File: VERSIONING.md
Comment (znewman01) [reply VERSIONING.md]: yeah forgot the link oops

File: VERSIONING.md
Comment (znewman01) [reply VERSIONING.md]: nice

File: VERSIONING.md
Comment (znewman01) [reply VERSIONING.md:39]: idk i like the footnote links

File: VERSIONING.md
Comment (znewman01) [reply VERSIONING.md:81]: changed header to "feature stability"

File: VERSIONING.md
Comment (znewman01) [reply VERSIONING.md]: This is actually a reasonable thing to do because we'll have a "deprecated behavior library" that all deprecated code imports. Added a note

**Comment by znewman01** [comment]
Rebased, addressed outstanding comments

**Review by loosebazooka — DISMISSED** [review]

**Review by priyawadhwa — APPROVED** [review]

File: VERSIONING.md
Comment (dekkagaijin) [inline VERSIONING.md:23]: As long as Cosign is going to cosplay as semver, it should be semver. If semver guarantees aren't going to be respected, then there should be an unambiguously (programmatically) different scheme.

File: VERSIONING.md
Comment (dekkagaijin) [inline VERSIONING.md:37]:
```suggestion
**Patch version increases.** These are "bug fix" releases, and should always be adopted.
```

File: VERSIONING.md
Comment (dekkagaijin) [inline VERSIONING.md:16]: `Better to promise a weaker guarantee and stick to it consistently`

? :p

Comment (znewman01) [reply VERSIONING.md:16]: > ? :p

? :p

Not sure what the source of confusion is here. I meant to say something along the lines of "before, we had a stronger versioning guarantee but didn't actually follow it; I'd rather have a policy we intend to follow."

Is there a better way to word that?

File: VERSIONING.md
Comment (znewman01) [reply VERSIONING.md]: SGTM, added

File: VERSIONING.md
Comment (znewman01) [reply VERSIONING.md:23]: Not quite sure what you mean by "cosplay as semver." Where do you think the confusion will come from? The MAJOR.MINOR.PATCH scheme? IMO semver doesn't have a monopoly on that. Or just the fact that we have pretty-similar-but-not-quite-the-same guarantees?

I'd be open to other versioning schemes as long as they had he following features:

- support for big, all-compatibility-bets-are-off breaking version bumps
- version bumps corresponding to "typical" releases which can be breaking but only in accordance with a deprecation policy
- releases that will always be safe to upgrade to, and that we may backport fixes to

Seems like a 3-part scheme is pretty natural given those goals. We could try to do something with dates I guess? `v2.20230111` but then the date in a patch release if we're going to backport something is awkward. WDYT?

**Review by priyawadhwa — APPROVED** [review]

**Comment by znewman01** [comment]
Alright, I'm going to merge this as it's been sitting for a month and a half now and represents more of a codification of what we've been doing than a real departure.

If there are things we'd like to refine, please feel free to chime in on this PR, the attached issue #2518 (which will remain open for a little while), or file a new issue. For minor things (e.g. typos, wordsmithing) feel free to send a PR.

**Pull request merged**
